apps/labs/posts/multi-dimensional-sparse-arrays.md

```diff
@@ -0,0 +1,225 @@
+
+Example:
+
+Let `A` have shape `(4, 5, 1, 3, 6)` and `B` have shape `(1, 9, 6, 7)`. The product's shape will end in `(3, 7)` (as `(m,n) × (n, k) -> (m, k)`). We then compare the leading dimensions of `A` and `B`, `(4, 5, 1)` and `(1, 9)` and expand the dimensions such that both arrays have the same leading dimensions: `(4, 5, 9)`. Now, we can broadcast `A` from shape `(4, 5, 1, 3, 6)` to `(4, 5, 9, 3, 6)`, and `B` from shape `(1, 9, 6, 7)` to `(4, 5, 9, 6, 7)`.
+
+Developing functionality for `matmul` involved considering two cases:
+
+1. **Multiplication of a sparse array by a sparse array** -
+   In this case, we first broadcast both arrays, then convert them to block-diagonal form in COO format (using helper function `_block_diag`). Afterward, we convert the 2D block-diagonal arrays into CSR format and use `csr_matmat` for efficient 2D sparse-sparse matrix multiplication, and convert the 2-D block diagonal product to an n-D COO array (using `_extract_block_diag`) to obtain the final result.
+
+   The process may be better understood through the following code:
+
+   ```python
+   # Determine the new shape to broadcast A and B
+   broadcast_shape = np.broadcast_shapes(shape_A[:-2], shape_B[:-2])
+   new_shape_A = broadcast_shape + shape_A[-2:]
+   new_shape_B = broadcast_shape + shape_B[-2:]
+
+   A_broadcasted = A.broadcast_to(new_shape_A)
+   B_broadcasted = B.broadcast_to(new_shape_B)
+
+   # Convert n-D COO arrays to 2-D block diagonal arrays
+   A_block_diag = _block_diag(A_broadcasted)
+   B_block_diag = _block_diag(B_broadcasted)
+
+   # Use csr_matmat to perform sparse matrix multiplication
+   C_block_diag = (A_block_diag @ B_block_diag).tocoo() # calls 2-D COO matmul, which routes via CSR
+
+   product_shape = broadcast_shape + (A.shape[-2], B.shape[-1])
+
+   # Convert the 2-D block diagonal array back to n-D
+   C = _extract_block_diag(C_block_diag, shape=product_shape)
+
+   return C
+   ```
+
+2. **Multiplication of a sparse array by a dense array** -
+   First, we broadcast `A` and `B`, then we call a C++ function `coo_matmat_dense_nd` which performs n-D sparse-dense multiplication (this is the n-D extension of the 2-D function `coo_matmat_dense` which I added in [PR #21240](https://github.com/scipy/scipy/pull/21240)).
+
+   Without diving too deep into the C++ logic, I want to introduce the concept of strides with the following code snippet:
+
+   ```cpp
+   npy_int64 strides_B[n_dim];
+   strides_B[n_dim - 1] = 1;
+   for (npy_int64 i = n_dim - 2; i >= 0; --i) {
+        strides_B[i] = strides_B[i + 1] * shape_B[i + 1];
+   }
+   ```
+
+   This code creates an array `strides`, which stores the step size needed to move from one element to the next along a particular axis in a multi-dimensional array.
```

Comment:
```suggestion
   This code creates an array `strides_B`, which stores the step size needed to move from one element to the next along a particular axis in a multi-dimensional array.
```

Reply:
I understand what you want to show, but I think this code may be overcomplicating things as you don't define `shape_B`, for example. Maybe just showing the formula can be enough?

Reply:
Would it make sense to generalise the code by removing the B in names like here - 2a8c9cadc71fff5e21d8ba916b4065028f2850cf? (please ignore the incorrect var name in the first line, I've corrected that in the next commit!)

And I could add a one-liner describing what shape means, something like "where `shape` is a tuple of dimension sizes of the array."

Reply:
Sure, that makes it simpler.

Reply:
Done!